### PR TITLE
Enable nullable reference types in some base classes

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Enable nullable reference types in some base classes in:
- ConsoleHostRawUserInterface.cs 
- InternalHostRawUserInterface.cs
- ServerRemoteHostRawUserInterface.cs
- MshHostRawUserInterface.cs
- regex.cs - Here Init() method refactored to return delegate and ensure that _isMatch always have a value.
- CommandProcessorBase.cs
- SessionStateScopeAPIs.cs

## PR Context

C# 8.0 introduces great feature nullable reference types.
https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/

The feature makes code more readable, understandable and reliable, it forces us to write cleaner code.

I started with some base classes so that using classes can benefit from this. 
I hope we will continue to roll up the feature after the PR.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by iSazonov** [comment]
@PowerShellTeam @vexx32  Could you please review the PR?

In the last days I have tried nullable reference types in different files and sometimes it gives an surprising result. I have already found several places where NRE could potentially be.
Also I found it is very useful for design understanding.

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 1]
```diff
@@ -63,7 +55,7 @@ internal CommandProcessorBase(CommandInfo commandInfo)
 
         #region properties
 
-        private InternalCommand _command;
+        private InternalCommand _command = default!;
```
Comment (vexx32) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs]: We're using the null-forgiving operator `!` quite a bit in this file. Is there a reason we prefer this over marking the properties as nullable?

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 2]
```diff
@@ -4,9 +4,8 @@
 using System.Collections;
 using System.Collections.ObjectModel;
 using System.Management.Automation.Internal;
-using System.Management.Automation.Language;
 
-using Dbg = System.Management.Automation.Diagnostics;
+#nullable enable
```
Comment (vexx32) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs]: I can already tell I'm not going to enjoy merging this into my Dispose{} branch for #9900 😂

File: src/System.Management.Automation/engine/SessionStateScopeAPIs.cs [hunk 3]
```diff
@@ -175,15 +175,17 @@ internal SessionStateScope CurrentScope
...
             set
             {
-                Diagnostics.Assert(
-                    value != null,
-                    "A null scope should never be set");
+                if (value == null)
+                {
+                     throw PSTraceSource.NewArgumentException(nameof(CurrentScope));
+                }
+
```
Comment (vexx32) [inline src/System.Management.Automation/engine/SessionStateScopeAPIs.cs]: Do we have an official guideline on where it is more appropriate to use `Diagnostics.Assert()` and where it's more appropriate to `throw`? I'd be hesitant to change them over without good reason.

**Review by vexx32 — APPROVED** [review]
Overall looks good, a few questions on some of the changes.

It might be worth putting some concrete guidelines together (e.g., in contributing docs) for refactoring in this manner and generally designing code that is going to use the non-nullable reference types feature set. That way we have a reference point for refactoring, reviewing, and developing, and we have a better chance at creating consistent implementation throughout the repo.

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 4]
```diff
@@ -18,13 +17,6 @@ internal abstract class CommandProcessorBase : IDisposable
     {
         #region ctor
 
-        /// <summary>
-        /// Default constructor.
-        /// </summary>
-        internal CommandProcessorBase()
-        {
-        }
-
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandProcessorBase.cs]: Yes, it is internal and it is safe to remove.

File: src/System.Management.Automation/engine/SessionStateScopeAPIs.cs (hunk 3 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/SessionStateScopeAPIs.cs]: We run all tests  (CIs) only on Release configuration. So Diagnostics.Assert() is not so useful.

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 5]
```diff
@@ -126,16 +118,22 @@ internal InternalCommand Command
             set
             {
                 // The command runtime needs to be set up...
-                if (value != null)
+                if (value == null)
+                {
+                    throw PSTraceSource.NewArgumentNullException(nameof(Command));
```
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandProcessorBase.cs]: I used follow trick - look references to the `Command`. We never check the property on null (there are some exclusions in exceptions). So design is that the property is non-nullable and value is never null.
And ... we should remove the check at all :-)

File: src/System.Management.Automation/engine/CommandProcessorBase.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandProcessorBase.cs]: It is design question. If a design intention is to disable null the implementation should follow the design. Review just should check the both.

File: src/System.Management.Automation/engine/CommandProcessorBase.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (iSazonov) [reply src/System.Management.Automation/engine/CommandProcessorBase.cs]: I added a comment to remove the check after we enable null reference types for all code base (SMA).

**Comment by xtqqczze** [comment]
@iSazonov What are the next steps on this PR? I would like start work on nullable reference types in other classes but we need support in the base classes first.

**Comment by iSazonov** [comment]
>  What are the next steps on this PR?

1. I hope MSFT team start the work and review the PR /cc @SteveL-MSFT 
2. Process all APIs starting with base classes.

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs [hunk 6]
```diff
@@ -17,6 +17,8 @@
 using WORD = System.UInt16;
```
Comment (PoshChan) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs]: @iSazonov, your last commit had 2 failures in `PowerShell-CI-macos`
Test-Connection.Ping.Force IPv4 with implicit PingOptions
```powershell
Expected exactly '10.79.2.17', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.165.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 84
84:             $result[0].Address | Should -BeExactly $realAddress
```
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.2.17', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.165.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 258
258:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

File: src/System.Management.Automation/namespaces/LocationGlobber.cs [hunk 7]
```diff
@@ -4698,7 +4698,7 @@ private static void TraceFilters(CmdletProviderContext context)
                 // Trace the filter
                 s_pathResolutionTracer.WriteLine("Filter: {0}", context.Filter ?? string.Empty);
 
-                if (context.Include != null)
+                if (context.Include != null && context.Include.Count > 0)
```
Comment (powercode) [inline src/System.Management.Automation/namespaces/LocationGlobber.cs:4701]: Maybe `if (context?.Include.Count > 0)`

File: src/System.Management.Automation/namespaces/LocationGlobber.cs [hunk 8]
```diff
@@ -4710,7 +4710,7 @@ private static void TraceFilters(CmdletProviderContext context)
                     s_pathResolutionTracer.WriteLine("Include: {0}", includeString.ToString());
                 }
 
-                if (context.Exclude != null)
+                if (context.Exclude != null && context.Exclude.Count > 0)
```
Comment (powercode) [inline src/System.Management.Automation/namespaces/LocationGlobber.cs:4713]: Maybe if (context?.Exclude.Count > 0)

**Review by powercode — APPROVED** [review]

File: src/System.Management.Automation/namespaces/LocationGlobber.cs (hunk 7 above)
Comment (vexx32) [reply src/System.Management.Automation/namespaces/LocationGlobber.cs:4701]: It looks like `if (context.Include?.Count > 0)` would be closer to the original framing here.

File: src/System.Management.Automation/namespaces/LocationGlobber.cs (hunk 8 above)
Comment (vexx32) [reply src/System.Management.Automation/namespaces/LocationGlobber.cs:4713]: Same as above :)

**Pull request closed without merging**